### PR TITLE
WIP: Optimize for big repository on repo home page to make it load less than 1s and /main less than 8s

### PR DESCRIPTION
**Pull request by lunny**
I tested it locally. The recent linux repo home page loads time from 8s to 2s if enabled memory last commit cache, 8s to 2.8s if enabled boltdb last commit cache.

**update**: now it only 1.9s even boltdb on my MacOS since we only get simple commit infomation.

**update(2019-02-22)**: added ls tree cache and files limit to 1000, now aport will /main directory will load less than 8s on my Macbook pro. I will add a log cache and need more work to clean up the codes.

blocked by https://github.com/go-gitea/git/pull/145

### CODE REVIEW

**Comment by lafriks** [comment]
Why not just reuse already existing caching functionality like commit counts are cached?

**Comment by lunny** [comment]
I think that cache infrastructures are too complicated. and I think the commit counting cache should also be used a separated cache service to do that.

**Comment by clandmeter** [comment]
Thanks a lot @lunny for trying to improve this usability concern.

I can confirm that using boltdb improves performance the most for me.
```
memory:
	initial load 120 seconds
	second load 60-70 seconds

boltdb:
	second load 40-50 seconds

none:
	initial load 120 seconds
	second load 130 seconds
```

url to try is http://51.15.253.24:3000/clandmeter/aports/src/branch/master/main

What really would make a difference on repo layouts like this would be a pager with a commit list limit set in app.ini. Or if that's not possible a hard limit like github does.

**Comment by lunny** [comment]
@clandmeter have you enabled `[cache]` on your app.ini. That will also help on commit counting. We have more improvements work to do. Maybe we should follow how github did that. limit 1000 files and hide all the last commit message?

**Review by kolaente — COMMENTED** [review]
Do you have another pr for `code.gitea.io/git`?

**Comment by clandmeter** [comment]
@lunny i enabled cache (redis) as this was a vanilla setup. the load is now around 30-40 seconds.

Yes a hard limit could be an option.
If the limit amount could be a setting users would be able to fine tune it based on system spec.

Github does the same: https://github.com/alpinelinux/aports/tree/master/main

This loads in a few seconds 1-3. I don't think gitea can do that even with a 1k limit. The amount of objects is now around 2k so i would assume at best it would load in 15 seconds.

Even with caching I see a lot of git commands (git rev-list -1 xxx) beeing executed in the background. I guess the cache doesn't cover all of the git lookups.

**Comment by lunny** [comment]
@clandmeter yes. I need send several PRs for all the optimizations since not all the git command be cached. I will send another PR to limit the display files number according user setting.

**Comment by clandmeter** [comment]
@lunny I wonder why you choose to use boltdb but did not add a Redis option? Redis is already used for caching and would provide much better scaling when your instance would grow?

**Comment by lafriks** [comment]
@lunny imho it would be better to implement type that uses internally existing caching methods. This way we would support also redis etc

**Comment by lunny** [comment]
@clandmeter It's easy to add redis.
@lafriks Will consider that.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 1]
```diff
@@ -317,13 +317,19 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
...
 - `MIGRATE`: **600**: Migrate external repositories timeout seconds.
 - `MIRROR`: **300**: Mirror external repositories timeout seconds.
 - `CLONE`: **300**: Git clone from internal repositories timeout seconds.
 - `PULL`: **300**: Git pull from internal repositories timeout seconds.
 - `GC`: **60**: Git repository GC timeout seconds.
 
+## Git - LastCommitCache settings (`git.last_commit_cache`)
+
+- `TYPE`: **none**: Cache type, could be empty, `memory` or `boltdb`.
+- `DATA_PATH`: ****: Cache dir when type is boltdb.
```
Comment (silverwind) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]: Can we have a default path?

**Comment by lunny** [comment]
@lafriks added old cache configuration support.

**Comment by lunny** [comment]
@clandmeter added redis support and reduced half of git commands when list last commit information.

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md (hunk 1 above)
Comment (lunny) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md]: The default type is none means do not use cache. For small repository, it's enough. So the default path is empty.

**Comment by clandmeter** [comment]
@lunny nice thx.
What is the config option to enable redis for commit cache?
Also applying this PR on master will segfault when using boltdb.

```
[Macaron] 2019-02-16 23:29:38: Started GET /clandmeter/aports/src/branch/master/main for 172.16.254.10
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x55cf09b37a95]

goroutine 1043 [running]:
code.gitea.io/gitea/modules/cache/lastcommit.(*BoltDBCache).Get(0x0, 0xc001a12540, 0x31, 0xc001a23b90, 0x28, 0xc0036648d0, 0x9, 0xc000513708, 0x55cf08d95855, 0xc000a35c80)
        /home/clandmeter/go/src/code.gitea.io/gitea/modules/cache/lastcommit/boltdb.go:57 +0xe5
code.gitea.io/gitea/vendor/code.gitea.io/git.targetedSearch(0xc000c86280, 0xc0024263c0, 0x55cf0a7f7860, 0x0)
        /home/clandmeter/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/commit_info.go:105 +0x5aa
created by code.gitea.io/gitea/vendor/code.gitea.io/git.getCommitsInfo
        /home/clandmeter/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/commit_info.go:247 +0x29c

```

**Comment by lunny** [comment]
Sorry, I changed the config option `DATA_PATH` to `CONN_STR`.

**Comment by clandmeter** [comment]
@lunny could you update the ini sample on how to use the new options? I am unable to make redis work. I can only make memory work.

File: modules/cache/ls_tree/cache.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package ls_tree
```
Comment (techknowlogick) [inline modules/cache/ls_tree/cache.go]: Copyright title please

File: modules/cache/ls_tree/init.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package ls_tree
```
Comment (techknowlogick) [inline modules/cache/ls_tree/init.go]: Same for this file

**Comment by lunny** [comment]
@clandmeter I send another commit today and I think that will make aport `/main` loading less than 8s(max 1000 files). And I will add another cache soon.

**Comment by clandmeter** [comment]
@lunny im not able to make this work anymore.
What is the correct way to add redis caching?
I used:

```
; Last commit cache
[git.last_commit_cache]
USE_DEFAULT_CACHE = true

[cache]
ADAPTER = redis
HOST = network=tcp,addr=:6379,db=0,pool_size=100,idle_timeout=180
```

which results in

```
[Macaron] 2019-02-22 17:02:14: Started GET /alpine/aports for 172.16.254.10
panic: interface conversion: interface {} is string, not *git.Commit

goroutine 1049 [running]:
code.gitea.io/gitea/modules/cache/lastcommit.(*lastCommitCache).Get(0xc00031a9e0, 0xc001989b00, 0x2d, 0xc001a2b5c0, 0x28, 0xc003a3f4a0, 0x9, 0x0, 0x0, 0x0)
        /home/clandmeter/go/src/code.gitea.io/gitea/modules/cache/lastcommit/cache.go:23 +0x103
code.gitea.io/gitea/vendor/code.gitea.io/git.targetedSearch(0xc00003b630, 0xc003a55620, 0x563ca2460060, 0xc00031a9e0)
        /home/clandmeter/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/commit_info.go:105 +0x5aa
created by code.gitea.io/gitea/vendor/code.gitea.io/git.getCommitsInfo
        /home/clandmeter/go/src/code.gitea.io/gitea/vendor/code.gitea.io/git/commit_info.go:247 +0x29c
```

**Comment by typeless** [comment]
I am wondering if [gitbase](https://github.com/src-d/gitbase) could be helpful here.
That way, we can use xorm on top of it and freeride on xorm's caching mechanism or other existing caching library for databases. 

I haven't carefully looked into gitbase, so this might be a naive idea. 😆

**Comment by lunny** [comment]
@typeless I don't know if gitbase is useful and related. For this PR, it should resolve all big repo home page issue. On my test, aports will only spent 800ms after throughout cached. But now there are some confusing issues I will resolve.

**Comment by clandmeter** [comment]
@lunny my observations with current code:

Using cache in memory gets page load to 500ms, impressive :)

Using boltdb is not working when using the same db for both git caches, gitea just wont start is this actually supported, if not maybe put that in the docs. If i use separate dbs than the result is as with redis.

Using Redis i dont see any improvements, page loads now in 15 seconds, before was 30 seconds but only 1000 items are loaded.

I would be nice if we could configure the 1000 limit or even disable it when caching is working properly.

config:

```
[cache]
ADAPTER = redis
HOST = network=tcp,addr=:6379,db=0,pool_size=100,idle_timeout=180

[cache.last_commit]
TYPE = redis
ENABLE_MIN_COMMIT_COUNT = 0
CONN_STR = addrs=127.0.0.1 password= db=0

[cache.ls_tree]
TYPE = redis
ENABLE_MIN_COMMIT_COUNT = 0
CONN_STR = addrs=127.0.0.1 password= db=0
```

**Comment by lunny** [comment]
@clandmeter Glad to hear that memory cache could be used. I have known that `boltdb` has some problems and I'm investigating that to fix it. I haven't tested `redis` so I don't know why it's slow. Of cousre, the limitation of show files will be added as a config item. I will continue the work. Thanks for your testing.

**Comment by clandmeter** [comment]
@lunny yes after retesting boltdb it seems broken.

Regarding redis, it looks like it doesn't cache anything. Compared to memory caching you can see the second run will do none or almost no git lookups anymore. With redis this keeps happening making it much slower. Funny thing is is does check for redis availability because if i turn it of it will return error 500.

I have some small remarks for your reference.

To my understanding a redis lookup should be much faster than a git lookup. So if you enable global cache with redis why not automatically enable it for the other cache options?

Also what does the following mean:

> repos which have less than commits number will enable this cache, 0 will always cache if TYPE is not empty

This reads as repo with more than 1000 commits (default setting) have caching disabled? 
The amount of commits in a tree has nothing to do with this caching option, or should i read this differently?

Thanks for looking into it, and ill keep following this PR.

**Comment by filipnavara** [comment]
After re-running my Bloom filter experiment from December based on [this article](https://devblogs.microsoft.com/devops/super-charging-the-git-commit-graph-iv-bloom-filters/) I think that's a better approach for solving the problem. It offers a way to accelerate the listing for any branch or history point. I realise the burden of proof is on me and unfortunately I may not be able to deliver it in timely manner. This may still be the way to go short-term but if someone decides to invest time into it there are viable alternatives. I'm now going through the discussion on the Git mailing list to read up on the various attempts to implement it and whether it got any traction since last year.

**Comment by filipnavara** [comment]
I still don't have much to share about my Bloom filter experiments in terms of code, but I do have some preliminary numbers from testing them on the Linux repository. Since this PR was started Gitea underwent some changes that changed some performance characteristics, mostly for the better. Migrating to `go-git` allowed me to experiment with caching some intermediate results that can efficiently help answer many different types of history queries such as the one used for repository listing.

This is all driven by the GitHub [enigineers philosophy](https://github.blog/2015-09-22-counting-objects/):
> Generally speaking, caching specific results to queries is a weak approach to
performance in complex systems. What you want to do instead is caching
intermediate steps of the computation, to be able to efficiently answer any kind
of query.

Listing the recent commits in Linux repository has few challenges:

1) To answer the question "What files are the root directory at HEAD revision and when were they changed last time?" it is necessary to traverse 8954 commits. This actually means loading 8954 commit objects and just as many tree objects from the repository.
2) The Linux repository is highly packed and these objects are often delta encoded so the actual number of objects that have to be loaded is higher than that.
3) `go-git` is not too efficient for huge pack files with delta encoded objects. Due to some implementation decisions about caching objects by their respective hash loading a delta encoded object using offset-delta encoding requires building a reverse offset-hash mapping that is quite expensive.

How can this be addressed?

- For most of the commit objects we don't actually need to load the whole object. We only care about the `tree hash` and the parent commits. Only for the last few commits we actually display we care about the commit message, autor and other properties. Turns out that this is already something tackled in regular Git by a precomputed `commit-graph` file. It stores minimal information needed for walking the commit graph in a form that is efficient for random access. In the case of Linux repository this file takes around 46 Mb (pack file is 2,5Gb + 200Mb index) for all commits in the repository in all branches. Using the `commit-graph` file saves loading nearly all of the commit objects. Work is underway to implement the support for this in `go-git`, but Gitea changes are needed to take full advantage of it (eg. scheduling when to rebuild the commit-graph files and executing it in the background).
- The number of loaded tree objects could be reduced by querying Bloom filters that can efficiently answer whether a certain path may have been changed in a given commit. With 99% probability (and 0% false negatives) the filters can correctly answer whether we need to load and compare two given commits or whether we can skip over to the parent commit straight away. In Linux repository this allows to skip loading trees for roughly 4738 of the 8954 commits (for comparison, in Rails repository it filters out 19432 of 22997 commits). The file size of the stored bloom filters is roughly 26 Mb.
- The `go-git` bottleneck with building the reverse offset-hash map for a packfile can be eliminated entirely. It's just a bit tricky to do so while keeping the external API shape intact. I've done some changes upstream that significantly improved the time to build this map, but it still takes a little over 1 second to do so.

Overall, with all the above performance enhancements I was able to get the load time of the repository listing page down to respectable 1.2 second.

**Comment by clandmeter** [comment]
@lunny any idea when this is going to be finalized/implemented?

**Comment by lunny** [comment]
@clandmeter will continue work on this PR. I wish this could be released in v1.9

**Comment by kolaente** [comment]
@lunny any updates on this?

**Comment by lunny** [comment]
This PR is outdate and I will send another PR to do some optimization.

**Comment by lunny** [comment]
Which gitea version are you using? @michaelshiel

**Comment by typeless** [comment]
@michaelshiel We have a repo that about 10 GB and it works fine. Does your repository have a large number of files at the root directory?

**Comment by lunny** [comment]
@typeless when you have many directories and many commits, it is still slow.

**Comment by michaelshiel** [comment]
Version: Gitea v1.10.0+dev-226-g256b17817 built with GNU Make 4.2.1, go1.12.9 : bindata, sqlite, sqlite_unlock_notify

10145 commits, 517 branches, and 1.9GB total repo size.

**Comment by michaelshiel** [comment]
To update, the fault appears to have been on my end. I initially was running the latest image in docker on OSX (with delegated volume mappings), running on Ubuntu 18.04 has no performance problems so far and any page load is sub 1 second. Sorry for the noise, and thanks again!

**Comment by kolaente** [comment]
@lunny any updates on this?

**Comment by lunny** [comment]
@kolaente I will start another PR after some time. This PR is too far away current master.

**Comment by lunny** [comment]
This has been replaced by #10069

**Pull request closed without merging**